spdlog/src/formatter/json_formatter.rs

```diff
@@ -0,0 +1,184 @@
+use std::fmt::{self, Write};
+
+use cfg_if::cfg_if;
+use serde::{ser::SerializeStruct, Serialize};
+
+use crate::{
+    formatter::{FmtExtraInfo, Formatter, LOCAL_TIME_CACHER},
+    Error, Record, StringBuf, __EOL,
+};
+
+struct JsonRecord<'a>(&'a Record<'a>);
+
+impl<'a> Serialize for JsonRecord<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let src_loc = self.0.source_location();
+
+        let mut record =
+            serializer.serialize_struct("JsonRecord", if src_loc.is_none() { 4 } else { 5 })?;
+
+        record.serialize_field("level", &self.0.level())?;
+        {
+            let mut local_time_cacher = LOCAL_TIME_CACHER.lock();
+            record.serialize_field::<str>(
+                "time",
+                local_time_cacher
+                    .get(self.0.time())
+                    .full_second_str()
+                    .as_ref(),
+            )?;
+        }
+        record.serialize_field("tid", &self.0.tid())?;
+        record.serialize_field("payload", self.0.payload())?;
+
+        if let Some(src_loc) = src_loc {
+            record.serialize_field("source_location", src_loc)?;
+        }
+
+        record.end()
+    }
+}
+
+impl<'a> From<&'a Record<'a>> for JsonRecord<'a> {
+    fn from(value: &'a Record<'a>) -> Self {
+        JsonRecord(value)
+    }
+}
+
+enum JsonFormatterError {
+    Fmt(fmt::Error),
+    Serialization(serde_json::Error),
+}
+
+impl From<fmt::Error> for JsonFormatterError {
+    fn from(value: fmt::Error) -> Self {
+        JsonFormatterError::Fmt(value)
+    }
+}
+
+impl From<serde_json::Error> for JsonFormatterError {
+    fn from(value: serde_json::Error) -> Self {
+        JsonFormatterError::Serialization(value)
+    }
+}
+
+impl From<JsonFormatterError> for crate::Error {
+    fn from(value: JsonFormatterError) -> Self {
+        match value {
+            JsonFormatterError::Fmt(e) => Error::FormatRecord(e),
+            JsonFormatterError::Serialization(e) => Error::Serialization(e.into()),
+        }
+    }
+}
+
+/// JSON logs formatter
+///
+/// Output format:
+///
+///   ```json
+///   {"level":"info","time":"2024-01-01
+/// 12:00:00","tid":123456,"payload":"test"}
+///
+///   // with source location
+///   {"level":"info","time":"2024-01-01
+/// 12:00:00","tid":123456,"payload":"test","source_location":{"module_path":"
+/// module","file":"file.rs","line":42}}
+///   ```
+#[derive(Clone)]
+pub struct JsonFormatter;
```

Comment:
```suggestion
#[derive(Clone)]
pub struct JsonFormatter(PhantomData<()>);
```

This is a public struct, writing in `pub struct T;` or `pub struct T {}` allows the user to bypass our method `T::new` to construct the struct directly. If we add private fields to it in the future, this causes it to become a breaking change.